src/bag/layout/template.py

```diff
@@ -2691,6 +2691,66 @@ def do_power_fill(self, layer_id: int, tr_manager: TrackManager,
+                            uniform_grid: bool = False) -> List[List[WireArray]]:
+        """Draw power fill on the given layer. Accepts as many different supply nets as provided.
+
+        Parameters
+        ----------
+        layer_id : int
+            the layer ID on which to draw power fill.
+        tr_manager : TrackManager
+            the TrackManager object.
+        sup_list : List[Union[WireArray, List[WireArray]]]
+            a list of supply wires to draw power fill for.
+        bound_box : Optional[BBox]
+            bound box over which to draw the power fill
+        x_margin : int
+            keepout margin on the x-axis. Fill is centered within margin.
+        y_margin : int
+            keepout margin on the y-axis. Fill is centered within margin.
+        uniform_grid : bool
+            draw power fill on a common grid instead of dense packing.
+
+        Returns
+        -------
+        List[List[WireArray]]
+            List of the wire arrays for each supply in sup_list, given in the
+            same order as sup_list.
+        """
+        if bound_box is None:
+            if self.bound_box is None:
+                raise ValueError("bound_box is not set")
+            bound_box = self.bound_box
+        bound_box = bound_box.expand(dx=-x_margin, dy=-y_margin)
+        is_horizontal = (self.grid.get_direction(layer_id) == 0)
+        if is_horizontal:
+            cl, cu = bound_box.yl, bound_box.yh
+            lower, upper = bound_box.xl, bound_box.xh
+        else:
+            cl, cu = bound_box.xl, bound_box.xh
+            lower, upper = bound_box.yl, bound_box.yh
+        fill_width = tr_manager.get_width(layer_id, 'sup')
+        fill_space = tr_manager.get_sep(layer_id, ('sup', 'sup'))
+        sep_margin = tr_manager.get_sep(layer_id, ('sup', ''))
+        tr_bot = self.grid.coord_to_track(layer_id, cl, mode=RoundMode.GREATER_EQ)
+        tr_top = self.grid.coord_to_track(layer_id, cu, mode=RoundMode.LESS_EQ)
+        trs = self.get_available_tracks(layer_id, tid_lo=tr_bot, tid_hi=tr_top, lower=lower, upper=upper,
+                                        width=fill_width, sep=fill_space, sep_margin=sep_margin,
+                                        uniform_grid=uniform_grid)
+        all_wars = [[] for _ in range(len(sup_list))]
+        htr_sep = HalfInt.convert(fill_space).dbl_value
+        if len(trs) < len(sup_list):
+            raise ValueError('Not enough available tracks to fill for all provided supplies')
+        for ncur, tr_idx in enumerate(trs):
+            warr = self.add_wires(layer_id, tr_idx, lower, upper, width=fill_width)
+            _ncur = HalfInt.convert(tr_idx).dbl_value // htr_sep if uniform_grid else ncur
+            all_wars[_ncur % len(sup_list)].append(warr)
+        for top_warr, bot_warr in zip(sup_list, all_wars):
+            self.draw_vias_on_intersections(top_warr, bot_warr)
+        return all_wars
+
```

Comment:
It will be good to rewrite the old do_power_fill() by passing appropriate derived parameters to this new method, so that the core algorithm is not repeated twice.